### PR TITLE
Add destination-scoped transaction replay protection

### LINKED ISSUE #740: RFC: Destination-scoped replay protection for signed transactions without trusting sender timestamps
## Summary

Define destination-scoped replay protection for signed Transactions without trusting the sender's claimed timestamp for ordering.

This is a protocol-design gate before changing the wire format. The previous proposal used one global hash chain per account. That does not compose with Rings traffic: an account can send to several destinations, each destination observes only a subset of the account's events, and Chord may drop or reorder messages. A gap in a global chain would therefore be normal rather than evidence of relay omission.

The mechanism in this issue complements the existing Session and MessageVerification lifetime checks. It does not replace ts_ms or ttl_ms: receiver time is still used to judge session authorization and message liveness.

## Target claim

For a retained stream state, the destination must be able to establish all of the following before handler dispatch:

- an exact signed transaction is dispatched at most once;
- a duplicate, a stale sequence, and two different signed transactions using the same sequence are distinct typed verdicts;
- the replay decision does not trust sender-provided wall-clock time;
- session-key rotation does not reset an account's stream;
- intermediate Chord relays remain stateless with respect to the origin's replay state.

The claim is receiver-local. It does not establish global order between accounts, reliable delivery, exactly-once side effects, or that a receiver which has lost all replay state can recognize an old message.

## Why the original global chain is not implementable

A frontier keyed only by sender DID assumes every receiver observes every event from that sender. That assumption is false under fan-out: if one account sends e0 to A and e1 to B, neither A nor B can extend a contiguous global frontier.

The previous loss rule was also insufficient. Advancing past a gap while retaining only (seq, hash) cannot later prove that a missing event belongs to the same chain. It also contradicts the stated no-omission and no-reordering laws.

Finally, replay protection survives restart only if the relevant state survives restart. Sender persistence alone is not enough; a receiver which forgets its frontier cannot distinguish a first delivery from a replay.

## Stream scope

The minimum stream identity is:

    StreamKey = (network_id, origin_account_did, destination_did)

The origin is the account DID recovered through Transaction.verification, not the delegated session DID and not the current relay. At one destination this reduces to one state record per origin account.

Any incarnation or reset identifier must be destination-authorized or cryptographically linked to the retained receiver frontier. A sender-controlled random stream id is not sufficient because it would let the sender abandon history at will.

Concurrent writers for one account and destination must either coordinate one durable sequence allocator or produce a typed fork. Multi-device and restored-backup behavior must be specified rather than treated as an ordinary restart.

## Selected state machine

Preserve Rings' best-effort delivery semantics with a bounded sequence window rather than a strict hash chain.

Each Transaction adds a sequence number inside the transaction signature. The receiver keeps a fixed-width window of accepted sequence slots and their transaction digests:

    SequenceState {
        high: u64,
        accepted: [Option<TransactionDigest>; W],
    }

The pure transition has typed outcomes:

    observe(state, seq, digest)
      -> First | Advance | Late | Replay | Fork | Stale

- First: no retained state exists; establish the first observed sequence as the baseline.
- Advance: seq is above high; move the fixed-size window without allocating in proportion to the gap, then record the digest.
- Late: seq is inside the retained window and its slot is empty; record and admit it.
- Replay: the slot already contains the same digest; reject it.
- Fork: the slot contains a different digest for the same sequence; reject it and return both signed transactions when both are retained.
- Stale: seq is below the retained window; reject it without dispatch.

This provides bounded out-of-order delivery and at-most-once dispatch while state is retained. It deliberately makes no no-omission or total-order claim. If strict in-order delivery is desired instead, the RFC must also specify retransmission, acknowledgements, a bounded durable outbox, gap recovery, and head-of-line blocking; adding only prev is not sufficient.

The transaction digest and signature transcript must use the new v2 domain and bind at least network_id, destination, tx_id, sequence, and data. Reuse the domain-separation discipline from #724. Sequence is mandatory in the new Transaction format; legacy transactions must not decode or dispatch after the cutover.

## Persistence and crash boundary

- Sender: reserve and persist the next sequence for StreamKey before exposing a signature. A crash may create a gap but must not reuse a sequence.
- Receiver: persist every state transition that admits a transaction before handler dispatch if the stated at-most-once claim is to survive process restart.
- Crash semantics: persistence-before-dispatch can lose an event if the process crashes between those steps. It prevents duplicate dispatch but is not exactly-once execution. Handlers are not assumed transactional with the replay store.
- Counter exhaustion: u64 exhaustion must fail closed or enter an explicit destination-authorized reset protocol; wrapping is forbidden.

A receiver that deletes a StreamKey record cannot preserve the replay claim for that stream. The design must therefore choose and document one of:

1. durable compact retention with an explicit storage bound and operator policy; or
2. a destination-issued fresh incarnation/challenge before an evicted stream may dispatch again.

Plain LRU eviction followed by accepting the next message as First is not security-preserving.

## Layering

- core: versioned Transaction transcript, StreamKey, SequenceState, TransactionDigest, and the pure observe reducer.
- final logical destination: run the reducer after both payload and transaction signatures verify and before application validation/handler effects are allowed to dispatch.
- relays: forward without creating or advancing origin replay state.
- storage adapters: persist sender allocators and receiver state on native and Wasm targets. This may require wiring outside core; do not assume node and RPC remain unchanged until the ownership boundary is demonstrated.
- observability: typed counters for Replay, Fork, Stale, persistence failure, and reset/incarnation failure.

## Protocol cutover (no compatibility)

Adding the required sequence field is a deliberate breaking change to both the Transaction wire format and its signature transcript. Roll it out as one explicit network-wide cutover:

- replace the legacy Transaction shape in place; sequence is required, never optional;
- bump the transaction signature domain from v1 to v2;
- update every seed, relay, destination, client, and Wasm consumer before the new network resumes;
- fail closed when legacy bytes cannot decode or verify.

There is no dual decoder, protocol negotiation, downgrade path, feature flag, or legacy fallback. Old and new nodes are not expected to interoperate. Release notes must identify the cutover release and state that mixed-version overlays are unsupported.

## Acceptance criteria

- [ ] Check in the protocol note describing the exact claim, StreamKey, state-retention assumption, crash boundary, and hard-cutover rule.
- [ ] Add pure model/property tests for duplicate rejection, bounded reordering, stale rejection, fork classification, sequence gaps, counter exhaustion, and deterministic transitions.
- [ ] Test that two destinations of one account advance independently without false gaps.
- [ ] Test that session-key rotation preserves the account/destination stream.
- [ ] Test sender restart without sequence reuse and receiver restart without replay acceptance.
- [ ] Demonstrate a hard bound for hot in-memory state and document how durable state is retained or safely re-incarnated.
- [ ] Run replay admission only at the final destination; relays remain stateless.
- [ ] Remove legacy Transaction decoding and prove old-format messages cannot reach dispatch after the cutover.
- [ ] Update SECURITY.md with the precise current guarantee and non-guarantees.
- [ ] Land #742 afterwards using the same per-origin destination table, without coupling quota verdicts to sequence verdicts.

## Non-goals

- A global clock or cross-account causal order.
- Reliable delivery or silently repairing missing transactions.
- Exactly-once application side effects.
- Treating a gap as proof that a particular relay dropped a message.
- Turning fork evidence into reputation or slashing policy; that belongs to the DRanking ledger.
- Reusing this state as the DRanking receipt DAG without a separate composition design.

Related: #742 (per-sender quota, follows this issue), #724 (signature domains), #736 (path-less relay), #711 (cross-account trust algebra).

### PR DESCRIPTION
**Pull request by zkjoie**
## Summary

- add the mandatory v2 destination-scoped transaction sequence and hard wire-format cutover
- persist bounded sender reservations before signing and receiver replay windows before final-destination dispatch
- keep intermediate relays and chunk-envelope framing outside logical replay state
- wire dedicated native and browser replay stores, document the security boundary, and bump the release to 0.24.0

## Validation

- `cargo +nightly-aarch64-apple-darwin fmt --all -- --check`
- `cargo clippy -p rings-core --all-targets --all-features -- -D warnings -D clippy::unwrap_used -D clippy::expect_used -D clippy::panic -D clippy::todo -D clippy::unimplemented -D clippy::indexing_slicing`
- `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- `cargo clippy -p rings-core --features wasm --no-deps --no-default-features --target=wasm32-unknown-unknown --tests -- -D warnings`
- `cargo test -p rings-core --all-targets --all-features --no-fail-fast -q`
- `cargo test -p rings-node --all-targets --all-features -q`
- `cargo check -p rings-core --target wasm32-unknown-unknown --no-default-features --features wasm`
- `cargo check -p rings-node --target wasm32-unknown-unknown --no-default-features --features browser`
- `cargo check --manifest-path frontend/Cargo.toml --target wasm32-unknown-unknown`

Closes #740

### CODE REVIEW

File: crates/core/src/swarm/mod.rs [hunk 1]
```diff
@@ -225,6 +229,14 @@ impl Swarm {
                 "Should be ConnectNodeSend".to_string(),
             ));
         };
+        if offer_payload.transaction.destination != self.did() {
+            return Err(Error::InvalidMessage(
+                "ConnectNodeSend destination does not match this node".to_string(),
+            ));
+        }
+        self.transport
+            .admit_transaction_replay(&offer_payload.transaction)
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/core/src/swarm/mod.rs:238]: Fixed in aabef761. A single named predicate now verifies the inner transaction and outer carrier, and answer_offer plus accept_answer invoke it before replay admission. The regression corrupts each inner signature, confirms rejection, then confirms the untouched payload still succeeds so invalid input cannot poison replay state.

File: crates/core/src/message/replay.rs [hunk 2]
```diff
@@ -0,0 +1,774 @@
...
+        let retained_min = state.retained_min();
+        return (state, SequenceVerdict::Stale { retained_min });
+    }
+    let verdict = state.observe_retained(sequence, digest);
+    (state, verdict)
+}
+
+/// Versioned durable sender and receiver state.
+///
+/// Fields are private so only [`TransactionReplay`] can apply the capacity and persistence laws.
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/core/src/message/replay.rs]: Fixed in aabef761. The private intra-doc link was removed, and RUSTDOCFLAGS=-Dwarnings cargo doc --all --no-deps now passes locally.

File: crates/core/src/message/replay.rs [hunk 3]
```diff
@@ -0,0 +1,774 @@
...
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct ReplaySnapshot {
+    sender: BTreeMap<StreamKey, u64>,
+    receiver: BTreeMap<StreamKey, SequenceState>,
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/core/src/message/replay.rs:244]: Fixed in aabef761. ReplaySnapshot now serializes as an opaque Rings-codec byte sequence before entering the JSON-compatible IndexedDB adapter, preserving structured keys and full-width counters. Added native encoding coverage and a real non-empty IndexedDB restart round trip using u64::MAX.

**Review by RyanKung — COMMENTED** [review]
LGTM
